src/libpmem2/pmem2_windows.c

```diff
@@ -35,52 +35,60 @@
  */
 
 #include "libpmem2.h"
+#include "out.h"
 #include "pmem2.h"
 #include "pmem2_utils.h"
 
+/*
+ * mapping -- creates or opens a named or unnamed file mapping object
+ *	for a specified file
+ */
+HANDLE
+mapping(const struct pmem2_config *cfg, DWORD protect, int *err)
+{
+	*err = 0;
+	HANDLE fh = CreateFileMapping(cfg->handle,
+		NULL, /* security attributes */
+		PAGE_READWRITE,
+		(DWORD)(0),
+		(DWORD)(0 & 0xFFFFFFFF),
+		NULL);
+
+	if (!(fh) || fh == INVALID_HANDLE_VALUE) {
+		*err = GetLastError();
+		ERR("CreateFileMapping, error: 0x%08x", err);
+	}
+
+	return fh;
+}
+
 /*
  * pmem2_map -- XXX
  */
 int
-pmem2_map(struct pmem2_config *cfg, struct pmem2_map **mapp)
+pmem2_map(const struct pmem2_config *cfg, struct pmem2_map **mapp)
 {
 	int ret = PMEM2_E_OK;
 	struct pmem2_map *map;
-	map = (struct pmem2_map *)pmem2_zalloc(&ret, sizeof(*map));
+	map = (struct pmem2_map *)pmem2_zalloc(sizeof(*map), &ret);
 	if (!map)
 		ret;
-	map->cfg = (struct pmem2_config *)pmem2_zalloc(&ret, sizeof(*map->cfg));
-	if (!map->cfg) {
+	map->cfg = (struct pmem2_config *)pmem2_zalloc(sizeof(*map->cfg), &ret);
+	if (!map->cfg)
 		goto err_free_map;
-	}
+
 	memcpy(map->cfg, cfg, sizeof(*cfg));
 	map->addr = NULL;
-	*mapp = map;
 
 	DWORD access = FILE_MAP_ALL_ACCESS;
-	size_t max_size = cfg->length + cfg->offset;
-	HANDLE fh = CreateFileMapping(cfg->handle,
-		NULL, /* security attributes */
-		PAGE_READWRITE,
-		(DWORD)(max_size >> 32),
-		(DWORD)(max_size & 0xFFFFFFFF),
-		NULL);
-
-	if (!(fh) || fh == INVALID_HANDLE_VALUE) {
-		ret = GetLastError();
-		if (ret == ERROR_ACCESS_DENIED) {
-			fh = CreateFileMapping(cfg->handle,
-				NULL, /* security attributes */
-				PAGE_READONLY,
-				(DWORD)(max_size >> 32),
-				(DWORD)(max_size & 0xFFFFFFFF),
-				NULL);
-			access = FILE_MAP_READ;
-		}
-		if (!(fh) || fh == INVALID_HANDLE_VALUE) {
-			ret = PMEM2_E_INVALID_HANDLE;
-			goto err_free_map_cfg;
+	HANDLE fh = mapping(map->cfg, PAGE_READWRITE, &ret);
+	if (ret == ERROR_ACCESS_DENIED) {
+		fh = mapping(map->cfg, PAGE_READONLY, &ret);
+		access = FILE_MAP_READ;
 		}
+	if (!(fh) || fh == INVALID_HANDLE_VALUE) {
+		ret = PMEM2_E_INVALID_HANDLE;
+		goto err_free_map_cfg;
```

Comment:
At this point, you do not have valid fh. err_free_map_cfg closes it. You need additional error case.